proxy/src/proxy/conntrack.rs

```diff
@@ -0,0 +1,680 @@
+use std::pin::Pin;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicU8, AtomicUsize, Ordering};
+use std::task::{Context, Poll};
+use std::time::SystemTime;
+use std::{fmt, io};
+
+use pin_project_lite::pin_project;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ConnId(usize);
+
+#[derive(Default)]
+pub struct ConnectionTracking {
+    conns: clashmap::ClashMap<ConnId, (ConnectionState, SystemTime)>,
```

Comment:
Random idea. I exposed a new internal type in clashmap: https://docs.rs/clashmap/latest/clashmap/struct.ClashCollection.html

It's nothing more than the sharded lock API.

I wonder if ClashCollection<SlotMap> would work as a more efficient map here. We generate a random number for the "hash", and the connection id is then that random number + slotmap key.

Or maybe we could use sharded_slab 🤷

This is likely fine though

Reply:
Interesting. I'll take a look, thanks! I'm not sure yet what `ConnectionTracking` will look like in the end. We may not want to keep a map of all connections, instead we directly feed the conn state changes into metrics and just update a per-conn idle timer.